### PR TITLE
Replace LiveData and callbacks with flows

### PR DESCRIPTION
**Pull request by fstanis**

### CODE REVIEW

File: app/src/main/java/com/devrel/android/minwos/data/networks/ConnectivityStatusListener.kt [hunk 1]
```diff
@@ -21,153 +21,165 @@ import android.net.LinkProperties
...
+    val flow: SharedFlow<ConnectivityStatus>
     fun refresh()
-    fun setCallback(callback: (ConnectivityStatus) -> Unit)
-    fun clearCallback()
-
-    override fun onStart(owner: LifecycleOwner) = startListening()
-    override fun onStop(owner: LifecycleOwner) = stopListening()
 }
 
 class ConnectivityStatusListenerImpl @Inject constructor(
```
Comment (pfmaggi) [inline app/src/main/java/com/devrel/android/minwos/data/networks/ConnectivityStatusListener.kt:45]: Please add a @ExperimentalCoroutinesApi

File: app/src/main/java/com/devrel/android/minwos/data/phonestate/TelephonyStatusListener.kt [hunk 2]
```diff
@@ -29,62 +29,61 @@ import android.telephony.SubscriptionManager.INVALID_SUBSCRIPTION_ID
...
     fun refresh()
-    fun setCallback(callback: (TelephonyStatus) -> Unit)
-    fun clearCallback()
-
-    override fun onStart(owner: LifecycleOwner) = startListening()
-    override fun onStop(owner: LifecycleOwner) = stopListening()
+    fun recheckPermissions()
 }
 
 class TelephonyStatusListenerImpl @Inject constructor(
```
Comment (pfmaggi) [inline app/src/main/java/com/devrel/android/minwos/data/phonestate/TelephonyStatusListener.kt:55]: Please add a @ExperimentalCoroutinesApi

File: app/src/main/java/com/devrel/android/minwos/data/util/RefreshFlow.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.collect
+
+/** A special type of hot flow that prevents "refresh spam" by dismissing refreshes emitted in a
+ *  short time period after the last accepted refresh.
+ */
+class RefreshFlow(
```
Comment (pfmaggi) [inline app/src/main/java/com/devrel/android/minwos/data/util/RefreshFlow.kt:29]: Do you need your own debounce implementation or can you use the standard one?
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/debounce.html

**Review by pfmaggi — APPROVED** [review]
LGTM, just a question regarding the `RefreshFlow`, can't you use the `debounce` in the standard lib?

**Comment by fstanis** [comment]
Unfortunately, `debounce` does a slightly different thing: most notably, `debounce(5s)` would wait 5s after an emit and then emit all subsequent emits in those 5s as a list, which means every tap on refresh is guaranteed to be delayed at least 5s. I think Kotlin/kotlinx.coroutines#1302 is what I'd need, but it's not implemented yet.

File: app/src/main/java/com/devrel/android/minwos/data/util/RefreshFlow.kt (hunk 3 above)
Comment (pfmaggi) [reply app/src/main/java/com/devrel/android/minwos/data/util/RefreshFlow.kt:29]: Thanks for the reply and the pointer to [Kotlin/kotlinx.coroutines#1302](https://github.com/Kotlin/kotlinx.coroutines/issues/1302).

**Pull request merged**